proxylib/redis/redisparser.go

```diff
@@ -0,0 +1,447 @@
+				log.Infof("RedisRule: key_regex mismatch '%v', '%s'", rule.keyRegexCompiled, key)
+				return false
+			}
+		} else {
+			log.Infof("Could not parse key for command '%s', rejecting", cmd)
+			return false
+		}
+	case cmdSpecialKey:
+		// TODO:  fixme to properly parse multi-key commands.
+		//        for now, just reject if there is a key regex specified
+		if len(regexStr) > 0 {
+			log.Warnf("Failed to match special-key cmd '%s' due to key-regex", cmd)
+			return false
+		}
+	case cmdNoKey:
+		break
+	}
+
+	return true
+}
+
+// RedisRuleParser parses protobuf L7 rules to enforcement objects
+// May panic
+func RedisRuleParser(rule *cilium.PortNetworkPolicyRule) []L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		ParseError("Can't get L7 rules.", rule)
+	}
+	var rules []L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var rr RedisRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "cmd":
+				rr.cmdExact = strings.ToLower(v)
+			case "key":
+				if v != "" {
+					rr.keyRegexCompiled = regexp.MustCompile(strings.ToLower(v))
+				}
+			default:
+				ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+
+		// ensure that 'cmd' value is a valid cmd and that key is
+		// only provided in scenarios when it makes sense.
+
+		if len(rr.cmdExact) > 0 {
+			// ensure this is a valid cmd
+			res := cmdMap[rr.cmdExact]
+			if res == cmdInvalid {
+				ParseError(fmt.Sprintf("Unable to parse L7 redis rule with invalid cmd: '%s'",
+					rr.cmdExact), rule)
+			} else if res == cmdNoKey && rr.keyRegexCompiled != nil {
+				ParseError(fmt.Sprintf("key regex not allowed for rule with cmd '%s'", rr.cmdExact), rule)
+			} else if res == cmdSpecialKey && rr.keyRegexCompiled != nil {
+				// this is temporary, as in the future we will properly parse a list of keys
+				// from these commands, and potentially allow a list of keys to match
+				ParseError(fmt.Sprintf("key regex not allowed for rule with cmd '%s'", rr.cmdExact), rule)
+			}
+
+		}
+
+		log.Debugf("Parsed RedisRule pair: %v", rr)
+		rules = append(rules, &rr)
+	}
+	return rules
+}
+
+type RedisParserFactory struct{}
+
+var redisParserFactory *RedisParserFactory
+
+func init() {
+	log.Info("init(): Registering redisParserFactory")
+	RegisterParserFactory("redis", redisParserFactory)
+	RegisterL7RuleParser("redis", RedisRuleParser)
+}
+
+type RedisParser struct {
+	connection *Connection
+	inserted   bool
+}
+
+/*
+type RedisRequestInfo struct {
+	cmd string
+	keys []string
+}
+*/
+
+const accessDeniedStr = "-Error Access Denied\r\n"
+
+func (pf *RedisParserFactory) Create(connection *Connection) Parser {
+	log.Debugf("RedisParserFactory: Create: %v", connection)
+
+	p := RedisParser{connection: connection}
+	return &p
+}
+
+func (p *RedisParser) OnData(reply, endStream bool, dataArray [][]byte) (OpType, int) {
+
+	// inefficient, but simple for now
+	data := bytes.Join(dataArray, []byte{})
+	if len(data) == 0 {
```

Comment:
This test can now be done before the `Join()` above. Also, I would use `MORE` only when a partial frame has been received:
```
if len(dataArray) == 0 {
    return NOP, 0
}
```